v2/blockstore/readwrite.go

```diff
@@ -59,48 +65,187 @@ func WithIndexPadding(p uint64) Option {
 	b.dedupCids = true
 }
 
-// NewReadWrite creates a new ReadWrite at the given path with a provided set of root CIDs as the car roots.
+// NewReadWrite creates a new ReadWrite at the given path with a provided set of root CIDs and options.
+//
+// ReadWrite.Finalize must be called once putting and reading blocks are no longer needed.
+// Upon calling ReadWrite.Finalize the CAR v2 header and index are written out onto the file and the
+// backing file is closed. Once finalized, all read and write calls to this blockstore will result
+// in panics. Note, a finalized file cannot be resumed from.
+//
+// If a file at given path does not exist, the instantiation will write car.Pragma and data payload
+// header (i.e. the inner CAR v1 header) onto the file before returning.
+//
+// When the given path already exists, the blockstore will attempt to resume from it.
+// On resumption the existing data frames in file are re-indexed, allowing the caller to continue
+// putting any remaining blocks without having to re-ingest blocks for which previous ReadWrite.Put
+// returned successfully.
+//
+// Resumption only works on files that were created by a previous instance of a ReadWrite
+// blockstore. This means a file created as a result of a successful call to NewReadWrite can be
+// resumed from as long as write operations such as ReadWrite.Put, and ReadWrite.PutMany returned
+// successfully and ReadWrite. On resumption the roots argument and WithCarV1Padding option must match the
+// previous instantiation of ReadWrite blockstore that created the file.
+// More explicitly, the file resuming from must:
+//   1. start with a complete CAR v2 car.Pragma.
+//   2. contain a complete CAR v1 data header with root CIDs matching the CIDs passed to the
+//      constructor, starting at offset optionally padded by WithCarV1Padding, followed by zero or
+//      more complete data frames. If any corrupt data frames are present the resumption will fail.
+//      Note, if set previously, the blockstore must use the same WithCarV1Padding option as before,
+//      since this option is used to locate the CAR v1 data payload.
+//  3. ReadWrite.Finalize must not have been called on the file.
 func NewReadWrite(path string, roots []cid.Cid, opts ...Option) (*ReadWrite, error) {
-	// TODO support resumption if the path provided contains partially written blocks in v2 format.
-	// TODO either lock the file or open exclusively; can we do somethign to reduce edge cases.
-	f, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0o666)
+	// Try and resume by default if the file exists.
+	resume := true
+	if _, err := os.Stat(path); err != nil { // TODO should we use stats to avoid resuming from files with zero size?
+		if os.IsNotExist(err) {
+			resume = false
+		} else {
+			return nil, err
+		}
+	}
+	f, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE, 0o666) // TODO: Should the user be able to configure FileMode permissions?
```

Comment:
@masih Open once then stat on the open file, because it is possible for the file to have moved in between stat and open calls.